File: packages/core/src/pipeline/events.ts
import type { StageEnum, StageOutput } from "../types/pipeline.js";
import type { DemandMetadata } from "../types/demand.js";
import type { GateResult } from "../types/gate.js";
import type { IssueClass } from "../types/issue.js";

interface EventBase {
  pipelineId: string;
  timestamp: Date;
}

export type PipelineEvent =
  | PipelineCreated
  | PipelineStarted
  | PipelineCompleted
  | PipelineFailed
  | PipelineCancelled
  | PipelineWaitingMerge
  | StageStarted
  | StageCompleted
  | StageFailed
  | StageBlocked
  | StageResumed
  | StageRegressed
  | AgentDispatched
  | AgentProgress
  | AgentCompleted
  | AgentFailed
  | GateEvaluated
  | GateRegressTriggered
  | HumanAssistRequested
  | HumanAssistReceived
  | HumanAssistEscalated
  | CredentialRefreshed
  | CredentialExpired
  | GitCommit
  | GitPush
  | MrCreated
  | MrMerged
  | WorktreeCreated
  | WorktreeCleaned
  | RollbackExecuted;

interface PipelineCreated extends EventBase { kind: "pipeline_created"; demandRef: DemandMetadata }
interface PipelineStarted extends EventBase { kind: "pipeline_started" }
interface PipelineCompleted extends EventBase { kind: "pipeline_completed" }
interface PipelineFailed extends EventBase { kind: "pipeline_failed"; reason: string }
interface PipelineCancelled extends EventBase { kind: "pipeline_cancelled"; operator: string }
interface PipelineWaitingMerge extends EventBase { kind: "pipeline_waiting_merge"; mrUrls: string[] }

interface StageStarted extends EventBase { kind: "stage_started"; stage: StageEnum }
interface StageCompleted extends EventBase { kind: "stage_completed"; stage: StageEnum; output: StageOutput }
interface StageFailed extends EventBase { kind: "stage_failed"; stage: StageEnum; error: string; retryCount: number }
interface StageBlocked extends EventBase { kind: "stage_blocked"; stage: StageEnum; issueClass: IssueClass }
interface StageResumed extends EventBase { kind: "stage_resumed"; stage: StageEnum; resolution: string }
interface StageRegressed extends EventBase { kind: "stage_regressed"; from: StageEnum; to: StageEnum; reason: string }

interface AgentDispatched extends EventBase { kind: "agent_dispatched"; stage: StageEnum; sessionId: string }
interface AgentProgress extends EventBase { kind: "agent_progress"; stage: StageEnum; message: string }
interface AgentCompleted extends EventBase { kind: "agent_completed"; stage: StageEnum; costUsd: number }
interface AgentFailed extends EventBase { kind: "agent_failed"; stage: StageEnum; error: string }

interface GateEvaluated extends EventBase { kind: "gate_evaluated"; stage: StageEnum; results: GateResult[] }
interface GateRegressTriggered extends EventBase { kind: "gate_regress_triggered"; rule: string; from: StageEnum; to: StageEnum }

interface HumanAssistRequested extends EventBase { kind: "human_assist_requested"; issueClass: IssueClass; target: string }
interface HumanAssistReceived extends EventBase { kind: "human_assist_received"; action: string; message: string }
interface HumanAssistEscalated extends EventBase { kind: "human_assist_escalated"; level: number }

interface CredentialRefreshed extends EventBase { kind: "credential_refreshed" }
interface CredentialExpired extends EventBase { kind: "credential_expired"; stage: StageEnum }

interface GitCommit extends EventBase { kind: "git_commit"; repo: string; hash: string }
interface GitPush extends EventBase { kind: "git_push"; repo: string; branch: string }
interface MrCreated extends EventBase { kind: "mr_created"; repo: string; url: string; iid: number }
interface MrMerged extends EventBase { kind: "mr_merged"; repo: string; url: string }
interface WorktreeCreated extends EventBase { kind: "worktree_created"; repo: string; path: string }
interface WorktreeCleaned extends EventBase { kind: "worktree_cleaned"; repo: string; reason: string }
interface RollbackExecuted extends EventBase { kind: "rollback_executed"; type: string; detail: string }

function base(pipelineId: string): EventBase {
  return { pipelineId, timestamp: new Date() };
}

export function pipelineCreatedEvent(pipelineId: string, demandRef: DemandMetadata): PipelineCreated {
  return { ...base(pipelineId), kind: "pipeline_created", demandRef };
}
export function pipelineStartedEvent(pipelineId: string): PipelineStarted {
  return { ...base(pipelineId), kind: "pipeline_started" };
}
export function pipelineCompletedEvent(pipelineId: string): PipelineCompleted {
  return { ...base(pipelineId), kind: "pipeline_completed" };
}
export function pipelineFailedEvent(pipelineId: string, reason: string): PipelineFailed {
  return { ...base(pipelineId), kind: "pipeline_failed", reason };
}
export function pipelineCancelledEvent(pipelineId: string, operator: string): PipelineCancelled {
  return { ...base(pipelineId), kind: "pipeline_cancelled", operator };
}
export function pipelineWaitingMergeEvent(pipelineId: string, mrUrls: string[]): PipelineWaitingMerge {
  return { ...base(pipelineId), kind: "pipeline_waiting_merge", mrUrls };
}

export function stageStartedEvent(pipelineId: string, stage: StageEnum): StageStarted {
  return { ...base(pipelineId), kind: "stage_started", stage };
}
export function stageCompletedEvent(pipelineId: string, stage: StageEnum, output: StageOutput): StageCompleted {
  return { ...base(pipelineId), kind: "stage_completed", stage, output };
}
export function stageFailedEvent(pipelineId: string, stage: StageEnum, error: string, retryCount: number): StageFailed {
  return { ...base(pipelineId), kind: "stage_failed", stage, error, retryCount };
}
export function stageBlockedEvent(pipelineId: string, stage: StageEnum, issueClass: IssueClass): StageBlocked {
  return { ...base(pipelineId), kind: "stage_blocked", stage, issueClass };
}
export function stageResumedEvent(pipelineId: string, stage: StageEnum, resolution: string): StageResumed {
  return { ...base(pipelineId), kind: "stage_resumed", stage, resolution };
}
export function stageRegressedEvent(pipelineId: string, from: StageEnum, to: StageEnum, reason: string): StageRegressed {
  return { ...base(pipelineId), kind: "stage_regressed", from, to, reason };
}

export function agentDispatchedEvent(pipelineId: string, stage: StageEnum, sessionId: string): AgentDispatched {
  return { ...base(pipelineId), kind: "agent_dispatched", stage, sessionId };
}
export function agentProgressEvent(pipelineId: string, stage: StageEnum, message: string): AgentProgress {
  return { ...base(pipelineId), kind: "agent_progress", stage, message };
}
export function agentCompletedEvent(pipelineId: string, stage: StageEnum, costUsd: number): AgentCompleted {
  return { ...base(pipelineId), kind: "agent_completed", stage, costUsd };
}
export function agentFailedEvent(pipelineId: string, stage: StageEnum, error: string): AgentFailed {
  return { ...base(pipelineId), kind: "agent_failed", stage, error };
}

export function gateEvaluatedEvent(pipelineId: string, stage: StageEnum, results: GateResult[]): GateEvaluated {
  return { ...base(pipelineId), kind: "gate_evaluated", stage, results };
}
export function gateRegressTriggeredEvent(pipelineId: string, rule: string, from: StageEnum, to: StageEnum): GateRegressTriggered {
  return { ...base(pipelineId), kind: "gate_regress_triggered", rule, from, to };
}

export function humanAssistRequestedEvent(pipelineId: string, issueClass: IssueClass, target: string): HumanAssistRequested {
  return { ...base(pipelineId), kind: "human_assist_requested", issueClass, target };
}
export function humanAssistReceivedEvent(pipelineId: string, action: string, message: string): HumanAssistReceived {
  return { ...base(pipelineId), kind: "human_assist_received", action, message };
}
export function humanAssistEscalatedEvent(pipelineId: string, level: number): HumanAssistEscalated {
  return { ...base(pipelineId), kind: "human_assist_escalated", level };
}

export function credentialRefreshedEvent(pipelineId: string): CredentialRefreshed {
  return { ...base(pipelineId), kind: "credential_refreshed" };
}
export function credentialExpiredEvent(pipelineId: string, stage: StageEnum): CredentialExpired {
  return { ...base(pipelineId), kind: "credential_expired", stage };
}

export function gitCommitEvent(pipelineId: string, repo: string, hash: string): GitCommit {
  return { ...base(pipelineId), kind: "git_commit", repo, hash };
}
export function gitPushEvent(pipelineId: string, repo: string, branch: string): GitPush {
  return { ...base(pipelineId), kind: "git_push", repo, branch };
}
export function mrCreatedEvent(pipelineId: string, repo: string, url: string, iid: number): MrCreated {
  return { ...base(pipelineId), kind: "mr_created", repo, url, iid };
}
export function mrMergedEvent(pipelineId: string, repo: string, url: string): MrMerged {
  return { ...base(pipelineId), kind: "mr_merged", repo, url };
}
export function worktreeCreatedEvent(pipelineId: string, repo: string, path: string): WorktreeCreated {
  return { ...base(pipelineId), kind: "worktree_created", repo, path };
}
export function worktreeCleanedEvent(pipelineId: string, repo: string, reason: string): WorktreeCleaned {
  return { ...base(pipelineId), kind: "worktree_cleaned", repo, reason };
}
export function rollbackExecutedEvent(pipelineId: string, type: string, detail: string): RollbackExecuted {
  return { ...base(pipelineId), kind: "rollback_executed", type, detail };
}

File: crates/poria-core/src/pipeline/events.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::types::{DemandMetadata, GateResult, IssueClass, StageEnum};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum PipelineEvent {
    PipelineCreated {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        demand_ref: Box<DemandMetadata>,
    },
    PipelineStarted {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
    },
    PipelineCompleted {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
    },
    PipelineFailed {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        reason: String,
    },
    PipelineCancelled {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        operator: String,
    },
    PipelineWaitingMerge {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        mr_urls: Vec<String>,
    },

    StageStarted {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        stage: StageEnum,
    },
    StageCompleted {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        stage: StageEnum,
        output: serde_json::Value,
    },
    StageFailed {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        stage: StageEnum,
        error: String,
        retry_count: i32,
    },
    StageBlocked {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        stage: StageEnum,
        issue_class: IssueClass,
    },
    StageResumed {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        stage: StageEnum,
        resolution: String,
    },
    StageRegressed {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        from: StageEnum,
        to: StageEnum,
        reason: String,
    },

    AgentDispatched {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        stage: StageEnum,
        session_id: String,
    },
    AgentProgress {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        stage: StageEnum,
        message: String,
    },
    AgentCompleted {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        stage: StageEnum,
        cost_usd: f64,
    },
    AgentFailed {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        stage: StageEnum,
        error: String,
    },

    GateEvaluated {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        stage: StageEnum,
        results: Vec<GateResult>,
    },
    GateRegressTriggered {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        rule: String,
        from: StageEnum,
        to: StageEnum,
    },

    HumanAssistRequested {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        issue_class: IssueClass,
        target: String,
    },
    HumanAssistReceived {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        action: String,
        message: String,
    },
    HumanAssistEscalated {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        level: i32,
    },

    CredentialRefreshed {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
    },
    CredentialExpired {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        stage: StageEnum,
    },

    GitCommit {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        repo: String,
        hash: String,
    },
    GitPush {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        repo: String,
        branch: String,
    },
    MrCreated {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        repo: String,
        url: String,
        iid: i64,
    },
    MrMerged {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        repo: String,
        url: String,
    },
    WorktreeCreated {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        repo: String,
        path: String,
    },
    WorktreeCleaned {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        repo: String,
        reason: String,
    },
    RollbackExecuted {
        pipeline_id: String,
        timestamp: DateTime<Utc>,
        #[serde(rename = "type")]
        rollback_type: String,
        detail: String,
    },
}

fn now() -> DateTime<Utc> {
    Utc::now()
}

impl PipelineEvent {
    pub fn pipeline_created(pipeline_id: impl Into<String>, demand_ref: DemandMetadata) -> Self {
        Self::PipelineCreated {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            demand_ref: Box::new(demand_ref),
        }
    }
    pub fn pipeline_started(pipeline_id: impl Into<String>) -> Self {
        Self::PipelineStarted {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
        }
    }
    pub fn pipeline_completed(pipeline_id: impl Into<String>) -> Self {
        Self::PipelineCompleted {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
        }
    }
    pub fn pipeline_failed(pipeline_id: impl Into<String>, reason: impl Into<String>) -> Self {
        Self::PipelineFailed {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            reason: reason.into(),
        }
    }
    pub fn pipeline_cancelled(pipeline_id: impl Into<String>, operator: impl Into<String>) -> Self {
        Self::PipelineCancelled {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            operator: operator.into(),
        }
    }
    pub fn pipeline_waiting_merge(pipeline_id: impl Into<String>, mr_urls: Vec<String>) -> Self {
        Self::PipelineWaitingMerge {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            mr_urls,
        }
    }

    pub fn stage_started(pipeline_id: impl Into<String>, stage: StageEnum) -> Self {
        Self::StageStarted {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            stage,
        }
    }
    pub fn stage_completed(
        pipeline_id: impl Into<String>,
        stage: StageEnum,
        output: serde_json::Value,
    ) -> Self {
        Self::StageCompleted {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            stage,
            output,
        }
    }
    pub fn stage_failed(
        pipeline_id: impl Into<String>,
        stage: StageEnum,
        error: impl Into<String>,
        retry_count: i32,
    ) -> Self {
        Self::StageFailed {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            stage,
            error: error.into(),
            retry_count,
        }
    }
    pub fn stage_blocked(
        pipeline_id: impl Into<String>,
        stage: StageEnum,
        issue_class: IssueClass,
    ) -> Self {
        Self::StageBlocked {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            stage,
            issue_class,
        }
    }
    pub fn stage_resumed(
        pipeline_id: impl Into<String>,
        stage: StageEnum,
        resolution: impl Into<String>,
    ) -> Self {
        Self::StageResumed {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            stage,
            resolution: resolution.into(),
        }
    }
    pub fn stage_regressed(
        pipeline_id: impl Into<String>,
        from: StageEnum,
        to: StageEnum,
        reason: impl Into<String>,
    ) -> Self {
        Self::StageRegressed {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            from,
            to,
            reason: reason.into(),
        }
    }

    pub fn agent_dispatched(
        pipeline_id: impl Into<String>,
        stage: StageEnum,
        session_id: impl Into<String>,
    ) -> Self {
        Self::AgentDispatched {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            stage,
            session_id: session_id.into(),
        }
    }
    pub fn agent_progress(
        pipeline_id: impl Into<String>,
        stage: StageEnum,
        message: impl Into<String>,
    ) -> Self {
        Self::AgentProgress {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            stage,
            message: message.into(),
        }
    }
    pub fn agent_completed(
        pipeline_id: impl Into<String>,
        stage: StageEnum,
        cost_usd: f64,
    ) -> Self {
        Self::AgentCompleted {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            stage,
            cost_usd,
        }
    }
    pub fn agent_failed(
        pipeline_id: impl Into<String>,
        stage: StageEnum,
        error: impl Into<String>,
    ) -> Self {
        Self::AgentFailed {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            stage,
            error: error.into(),
        }
    }

    pub fn gate_evaluated(
        pipeline_id: impl Into<String>,
        stage: StageEnum,
        results: Vec<GateResult>,
    ) -> Self {
        Self::GateEvaluated {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            stage,
            results,
        }
    }
    pub fn gate_regress_triggered(
        pipeline_id: impl Into<String>,
        rule: impl Into<String>,
        from: StageEnum,
        to: StageEnum,
    ) -> Self {
        Self::GateRegressTriggered {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            rule: rule.into(),
            from,
            to,
        }
    }

    pub fn human_assist_requested(
        pipeline_id: impl Into<String>,
        issue_class: IssueClass,
        target: impl Into<String>,
    ) -> Self {
        Self::HumanAssistRequested {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            issue_class,
            target: target.into(),
        }
    }
    pub fn human_assist_received(
        pipeline_id: impl Into<String>,
        action: impl Into<String>,
        message: impl Into<String>,
    ) -> Self {
        Self::HumanAssistReceived {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            action: action.into(),
            message: message.into(),
        }
    }
    pub fn human_assist_escalated(pipeline_id: impl Into<String>, level: i32) -> Self {
        Self::HumanAssistEscalated {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            level,
        }
    }

    pub fn credential_refreshed(pipeline_id: impl Into<String>) -> Self {
        Self::CredentialRefreshed {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
        }
    }
    pub fn credential_expired(pipeline_id: impl Into<String>, stage: StageEnum) -> Self {
        Self::CredentialExpired {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            stage,
        }
    }

    pub fn git_commit(
        pipeline_id: impl Into<String>,
        repo: impl Into<String>,
        hash: impl Into<String>,
    ) -> Self {
        Self::GitCommit {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            repo: repo.into(),
            hash: hash.into(),
        }
    }
    pub fn git_push(
        pipeline_id: impl Into<String>,
        repo: impl Into<String>,
        branch: impl Into<String>,
    ) -> Self {
        Self::GitPush {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            repo: repo.into(),
            branch: branch.into(),
        }
    }
    pub fn mr_created(
        pipeline_id: impl Into<String>,
        repo: impl Into<String>,
        url: impl Into<String>,
        iid: i64,
    ) -> Self {
        Self::MrCreated {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            repo: repo.into(),
            url: url.into(),
            iid,
        }
    }
    pub fn mr_merged(
        pipeline_id: impl Into<String>,
        repo: impl Into<String>,
        url: impl Into<String>,
    ) -> Self {
        Self::MrMerged {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            repo: repo.into(),
            url: url.into(),
        }
    }
    pub fn worktree_created(
        pipeline_id: impl Into<String>,
        repo: impl Into<String>,
        path: impl Into<String>,
    ) -> Self {
        Self::WorktreeCreated {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            repo: repo.into(),
            path: path.into(),
        }
    }
    pub fn worktree_cleaned(
        pipeline_id: impl Into<String>,
        repo: impl Into<String>,
        reason: impl Into<String>,
    ) -> Self {
        Self::WorktreeCleaned {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            repo: repo.into(),
            reason: reason.into(),
        }
    }
    pub fn rollback_executed(
        pipeline_id: impl Into<String>,
        rollback_type: impl Into<String>,
        detail: impl Into<String>,
    ) -> Self {
        Self::RollbackExecuted {
            pipeline_id: pipeline_id.into(),
            timestamp: now(),
            rollback_type: rollback_type.into(),
            detail: detail.into(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_event_serialization_tag() {
        let event = PipelineEvent::pipeline_started("pl-123");
        let json = serde_json::to_value(&event).unwrap();
        assert_eq!(json["kind"], "pipeline_started");
        assert_eq!(json["pipeline_id"], "pl-123");
        assert!(json["timestamp"].is_string());
    }

    #[test]
    fn test_stage_event_serialization() {
        let event = PipelineEvent::stage_started("pl-123", StageEnum::Dev);
        let json = serde_json::to_value(&event).unwrap();
        assert_eq!(json["kind"], "stage_started");
        assert_eq!(json["stage"], "dev");
    }

    #[test]
    fn test_event_roundtrip() {
        let event = PipelineEvent::pipeline_failed("pl-456", "timeout");
        let json_str = serde_json::to_string(&event).unwrap();
        let deserialized: PipelineEvent = serde_json::from_str(&json_str).unwrap();
        if let PipelineEvent::PipelineFailed { reason, .. } = deserialized {
            assert_eq!(reason, "timeout");
        } else {
            panic!("Wrong variant");
        }
    }
}
